packages/events-api/src/adapter.ts

```diff
@@ -0,0 +1,156 @@
+import EventEmitter from 'events'; // tslint:disable-line
+import http, { IncomingMessage, ServerResponse } from 'http';
+import debugFactory from 'debug'; // tslint:disable-line
+import { createHTTPHandler } from './http-handler';
+
+const debug = debugFactory('@slack/events-api:adapter');
+
+/**
+ * An adapter for Slack's Events API.
+ */
+export class SlackEventAdapter extends EventEmitter {
+  /**
+   * The token used to authenticate signed requests from Slack's Events API.
+   */
+  public readonly signingSecret: string;
```

Comment:
In `events-api` these fields are used in `http-handler`: `createHTTPHandler` accepts the adapter as an argument & accesses these options in a few different places (check for references therein to `adapter`).

I originally typed them as private, however given that they're accessed outside of their class implementation they can't really be "`private`".

There's two ways this *could*<sup>1</sup> be solved. Both involve passing the options that `createHTTPHandler` relies on through its arguments, but both vary on how that's done:

1. Modify the adapter's constructor to not destructure the opts it receives & store the `EventAdapterOptions` as a private field in `SlackEventAdapter`. These options can then be passed to `createHTTPHandler`.

```ts
class SlackEventAdapter {
  constructor(signingSecret: string, opts: EventAdapterOptions) {
    // ...
    this.opts = opts;
  }

  // ...

  public requestListener(): ... {
    return createHTTPHandler(this.signingSecret, this.opts);
  }
}
```

2. Recreate the opts arg (i.e. `{ includeBody: this.includeBody, includeHeaders: this.includeHeaders, ... }`) at the call site of `createHTTPHandler`.

```ts
public requestListener(): ... {
  return createHTTPHandler(this.signingSecret, {
    includeBody: this.includeBody,
    includeHeaders: this.includeHeaders,
    waitForResponse: this.waitForResponse,
  });
}
```

(1) is very tempting to do, however it creates a *very* odd situation. The previous fields (`includeBody`, `includeHeaders`, and `waitForResponse`) can't *really* be removed since somebody could easily be depending on reading them; however, if the pass-through options are stored in a different field, these legacy option fields are now suddenly completely unused 😬.

(2) is a bit tedious as any change in `EventAdapterOptions` needs to be propagated to `SlackEventAdapter#requestListener`, but it's overall the safer option.

Even more, both of these break a previous functionality: if a dev were to change the value of one of these options (i.e. `adapter.includeBody = false;`) *after* calling `requestListener`. Currently, the change would propagate to the HTTP handler, whereas these two solutions only record a single, static instant of these options. This *could* be remedied by making the options passed to `createHTTPHandler` dynamic:

```ts
return createHTTPHandler(() => this.signingSecret, {
  includeBody: () => this.includeBody,
  includeHeaders: () => this.includeHeaders,
  waitForResponse: () => this.waitForResponse,
});

// then in http-handler:
if (adapterOpts.includeBody()) {
  // ...
}

// or:
verifyRequestSignature({
  signingSecret: signingSecret(),
  // ...
});
```

but by this point we're *very* far down the rabbit hole.

Finally, my view on a JS -> TS conversion is that in JS-land (unless previously documented with `@private` or named with a leading `_`) everything is public--intentionally or not. That doesn't<sup>2</sup> hold back adding a `private` access modifier because a JS consumer won't see the difference after compiling--that is, they can still read the field from JS-land. The addition of `private` would only affect a JS user if they converted their codebase to TS (unlikely, but also filled with more roadbumps that this one would be relatively minor).

This doesn't translate to other API-level changes like a name or type change, however. Those kinds of changes propagate to the compiled JS output & therefore break JS code that could depend on the fields.

### tl;dr
The current impl. is very tied up and wasn't built on the same public/private access concerns.

---

<sup>1</sup> I haven't really analyzed these & they could easily break some minute piece of functionality.

<sup>2</sup> In the near future where [private class fields (a stage-3 proposal!)](https://github.com/tc39/proposal-class-fields) exist, TypeScript could theoretically compile `private x` to an `#x` field on classes, making adding `private` to a previously-untyped field a breaking change in the same way that changing its name would be a breaking change.

Reply:
thanks for the thorough discussion! i didn't notice that the properties were being accessed by the HTTP handler, so i thought it was unintentional to make these properties public, but you were being very intentional! 🙇 

> Finally, my view on a JS -> TS conversion is that in JS-land (unless previously documented with @private or named with a leading _) everything is public--intentionally or not. [...]

I agree with this point of view. As such I think the change here is perfectly fine. By the time we're ready to revise the public API, we may or may not need some or all of those options, so I think it would be a drain on effort to implement the dynamic getter suggestion just yet.